File: gams/apifiles/CSharp/xp_example2.cs
///////////////////////////////////////////////////////////////
// This program performs the following steps:
//    1. Generate a gdx file with demand data
//    2. Calls GAMS to solve a simple transportation model
//       (The GAMS model writes the solution to a gdx file)
//    3. The solution is read from the gdx file
///////////////////////////////////////////////////////////////

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace xp_example2
{
    class xp_example2
    {
        static gamsxcs gamsx;
        static gdxcs gdx;
        static optcs opt;

        static bool GDXError(int i, string s)
        {
            string msg = string.Empty;

            gdx.gdxErrorStr(i, ref msg);
            Console.WriteLine(s + "failed: " + msg);
            return false;
        }

        static bool WriteModelData(string fnGDXFile)
        {
            int status = 0;
            string msg = string.Empty;
            string[] Indx = new string[gamsglobals.maxdim];
            double[] Values = new double[gamsglobals.val_max];

            gdx.gdxOpenWrite(fnGDXFile, "XP_Example2", ref status);

            if (status != 0)
                return GDXError(status, "gdxOpenWrite");

            if (0 == gdx.gdxDataWriteStrStart("Demand", "Demand Data", 1, gamsglobals.dt_par, 0))
                return GDXError(gdx.gdxGetLastError(), "gdxDataWriteStrStart");

            Indx[0] = "New-York"; Values[gamsglobals.val_level] = 324.0; gdx.gdxDataWriteStr(Indx, Values);
            Indx[0] = "Chicago"; Values[gamsglobals.val_level] = 299.0; gdx.gdxDataWriteStr(Indx, Values);
            Indx[0] = "Topeka"; Values[gamsglobals.val_level] = 274.0; gdx.gdxDataWriteStr(Indx, Values);

            if (0 == gdx.gdxDataWriteDone())
                return GDXError(gdx.gdxGetLastError(), "gdxDataWriteDone");

            if (gdx.gdxClose() != 0)
                return GDXError(gdx.gdxGetLastError(), "gdxClose");

            return true;
        }

        static bool CallGams(string sysDir)
        {
            string msg = string.Empty;
            int saveEOLOnly;

            string defFile = sysDir + System.IO.Path.DirectorySeparatorChar + "optgams.def";

            if (opt.optReadDefinition(defFile) != 0)
            {
                int i, itype = 0;
                for (i = 1; i <= opt.optMessageCount(); i++)
                {
                    opt.optGetMessage(i, ref msg, ref itype);
                    Console.WriteLine(msg);
                }
                Console.WriteLine("*** Problem reading definition file " + defFile);
                return false;
            }

            saveEOLOnly = opt.optEOLOnlySet(0);
            //need to adjust path if executed e.g. in bin\debug            
            opt.optReadFromStr("I=..\\GAMS\\model2.gms lo=2");
            opt.optEOLOnlySet(saveEOLOnly);
            opt.optSetStrStr("sysdir", sysDir);

            if (gamsx.gamsxRunExecDLL(opt.GetoptPtr(), sysDir, 1, ref msg) != 0)
            {
                Console.WriteLine("Could not execute RunExecDLL: " + msg);
                return false;
            }
            return true;
        }

        static bool ReadSolutionData(string fnGDXFile)
        {
            int VarNr = 0, NrRecs = 0, FDim = 0, dim = 0, vartype = 0;
            int status = 0;
            string VarName;
            string msg = string.Empty;
            string[] Indx = new string[gamsglobals.maxdim];
            double[] Values = new double[gamsglobals.val_max];

            gdx.gdxOpenRead(fnGDXFile, ref status);
            if (status != 0)
                return GDXError(status, "gdxOpenRead");

            VarName = "result";
            if (0 == gdx.gdxFindSymbol(VarName, ref VarNr))
            {
                Console.WriteLine("Could not find variable >" + VarName + "<");
                return false;
            }

            gdx.gdxSymbolInfo(VarNr, ref VarName, ref dim, ref vartype);
            if (2 != dim || gamsglobals.dt_var != vartype)
            {
                Console.WriteLine(VarName + " is not a two dimensional variable");
                return false;
            }

            if (0 == gdx.gdxDataReadStrStart(VarNr, ref NrRecs))
                return GDXError(gdx.gdxGetLastError(), "gdxDataReadStrStart");

            while (gdx.gdxDataReadStr(ref Indx, ref Values, ref FDim) != 0)
            {
                int i;
                if (0.0 == Values[gamsglobals.val_level]) continue; /* skip level = 0.0 is default */
                for (i = 0; i < dim; i++)
                {
                    Console.Write(Indx[i]);
                    if (i < dim - 1)
                        Console.Write(".");
                }
                Console.WriteLine(" = " + Values[gamsglobals.val_level]);
            }
            Console.WriteLine("All solution values shown");

            gdx.gdxDataReadDone();

            status = gdx.gdxGetLastError();
            if (status != 0)
                return GDXError(status, "GDX");

            if (gdx.gdxClose() != 0)
                return GDXError(gdx.gdxGetLastError(), "gdxClose");

            return true;
        }



        static int Main(string[] args)
        {
            string sysDir, msg = string.Empty;
            string defSysDir = @"C:\GAMS\win32\24.0";

            if (Environment.GetCommandLineArgs().Length > 1)
                sysDir = Environment.GetCommandLineArgs()[1];
            else
                sysDir = defSysDir;

            Console.WriteLine("Loading objects from GAMS system directory: " + sysDir);

            gamsx = new gamsxcs(sysDir, ref msg);
            if (!string.IsNullOrEmpty(msg))
            {
                Console.WriteLine("Could not create gamsx object: " + msg);
                return 1;
            }
            gdx = new gdxcs(sysDir, ref msg);
            if (!string.IsNullOrEmpty(msg))
            {
                Console.WriteLine("Could not create gdx object: " + msg);
                return 1;
            }
            opt = new optcs(sysDir, ref msg);
            if (!string.IsNullOrEmpty(msg))
            {
                Console.WriteLine("Could not create opt object: " + msg);
                return 1;
            }

            if (!WriteModelData("demanddata.gdx"))
            {
                Console.WriteLine("Model data not written");
                return 1;
            }

            if (!CallGams(sysDir))
            {
                Console.WriteLine("Call to GAMS failed");
                return 1;
            }

            if (!ReadSolutionData("results.gdx"))
            {
                Console.WriteLine("Could not read solution back");
                return 1;
            }

            return 0;
        }
    }
}

File: gams/apifiles/Fortran/xp_example2.f90
! This program performs the following steps:
!    1. Generate a gdx file with demand data
!    2. Calls GAMS to solve a simple transportation model
!       (The GAMS model writes the solution to a gdx file)
!    3. The solution is read from the gdx file

MODULE exData
  USE gamsglobals
  IMPLICIT NONE
  CHARACTER(LEN=UEL_IDENT_LEN), DIMENSION(MAX_INDEX_DIM) :: Indx
  REAL(KIND=8), DIMENSION(val_max) :: Values
  INTEGER(KIND=8)    :: PGDX = 0
  INTEGER(KIND=8)    :: PGAX = 0
  INTEGER(KIND=8)    :: POPT = 0
END MODULE exData

PROGRAM xp_example2
  USE gamsxf9def
  USE gdxf9def
  USE optf9def
  USE gamsglobals
  USE exData
  IMPLICIT NONE
  
  LOGICAL			 :: ok
  INTEGER(KIND=4)    :: iargc
  INTEGER(KIND=4)    :: status
  CHARACTER(LEN=255) :: Msg, Sysdir

  IF (iargc() /= 1) THEN
    WRITE(*,*) '**** xp_Example2: incorrect number of parameters'
    CALL gdxExit(1)
  END IF

  CALL getarg(1, Sysdir)
  WRITE(*,*) 'Loading objects from GAMS system directory: ', Sysdir
  
! Create objects
  ok = gamsxCreateD(PGAX, Sysdir, Msg)
  IF (.NOT. ok) THEN
    WRITE(*,*) '**** Could not load GDX library'
    WRITE(*,*) '**** ', Msg
    CALL gdxExit(1)
  END IF

  ok = gdxCreateD(PGDX, Sysdir, Msg)
  IF (.NOT. ok) THEN
    WRITE(*,*) '**** Could not load GDX library'
    WRITE(*,*) '**** ', Msg
    CALL gdxExit(1)
  END IF

  ok = optCreateD(POPT, Sysdir, Msg)
  IF (.NOT. ok) THEN
    WRITE(*,*) '**** Could not load GDX library'
    WRITE(*,*) '**** ', Msg
    CALL gdxExit(1)
  END IF

  status = WriteModelData('demanddata.gdx')
  IF (status /= 0) THEN
    WRITE(*,*) 'Model data not written'
    CALL TERMINATE
  END IF

  status = CallGams(sysdir)
  IF (status /= 0) THEN
    WRITE(*,*) 'Call to GAMS failed'
    CALL TERMINATE
  END IF

  status = ReadSolutionData('results.gdx')
  IF (status /= 0) THEN
    WRITE(*,*) 'Could not read solution back'
    CALL TERMINATE
  END IF

CONTAINS
  INTEGER (KIND=4) FUNCTION WriteModelData(fngdxfile)
    USE exData
    CHARACTER(LEN=*) :: fngdxfile
    INTEGER(KIND=4)  :: tmp, status
  
    tmp = gdxOpenWrite(PGDX, fngdxfile, 'xp_Example2', status)
    IF (status /= 0) THEN
      CALL gdxerror(status, 'gdxOpenWrite')
    END IF

    status = gdxDataWriteStrStart(PGDX, 'Demand', 'Demand Data', 1, DT_PAR, 0)
    IF (status == 0) THEN
      CALL gdxerror(gdxGetLastError(PGDX), 'gdxDataWriteStrStart')
    END IF

!   Initalize some GDX data structure */
    Indx(1) = 'New-York'
    Values(VAL_LEVEL) = 324.0
    status = gdxDataWriteStr(PGDX, Indx, Values)
    Indx(1) = 'Chicago'
    Values(VAL_LEVEL) = 299.0
    status = gdxDataWriteStr(PGDX, Indx, Values)
    Indx(1) = 'Topeka'
    Values(VAL_LEVEL) = 274.0
    status = gdxDataWriteStr(PGDX, Indx, Values)

    status = gdxDataWriteDone(PGDX)
    IF (status == 0) THEN
      CALL gdxerror(gdxGetLastError(PGDX), 'gdxDataWriteDone')
    END IF

    status = gdxClose(PGDX)
    IF (status /= 0) THEN
      CALL gdxerror(gdxGetLastError(PGDX), 'gdxClose')
    END IF

    WriteModelData = 0
  END FUNCTION
 
  INTEGER (KIND=4) FUNCTION CallGams(sysdir)
    USE exData
    CHARACTER(LEN=*)   :: sysdir
    CHARACTER(LEN=255) :: deffile, msg
    INTEGER(KIND=4)    :: status, i, itype
    
    deffile = TRIM(sysdir) // '/optgams.def'
    status = optReadDefinition(POPT,deffile)
    IF (status /= 0) THEN
      DO i=1,optMessageCount(POPT)
        CALL optGetMessage(POPT, i, msg, itype)
        WRITE(*,*) msg
      END DO
      CallGams = 1
      RETURN
    END IF
    
    CALL optSetStrStr(POPT, 'SysDir',    sysDir)
    CALL optSetStrStr(POPT, 'Input',     '../GAMS/model2.gms')
    CALL optSetIntStr(POPT, 'LogOption', 3)    
    
    status = gamsxRunExecDLL(PGAX, POPT, sysdir, 1, msg)
    IF (status /= 0) THEN
      WRITE(*,*) 'Could not execute RunExecDLL:', msg
      CallGams = 1
      RETURN
    END IF

    CallGams = 0
  END FUNCTION
 


  INTEGER (KIND=4) FUNCTION ReadSolutionData(fngdxfile)
    USE exData
    CHARACTER(LEN=*)   :: fngdxfile
    CHARACTER(LEN=255) :: msg, VarName
    INTEGER(KIND=4)  :: status, tmp, VarNr, NrRecs, FDim, dim, VarType, D

    tmp = gdxOpenRead(PGDX, fngdxfile, status)
    IF (status /= 0) THEN
      CALL gdxerror(status, 'gdxOpenRead')
    END IF

    VarName = 'result'
    status = gdxFindSymbol(PGDX, VarName, VarNr)
    IF (status == 0) THEN
      WRITE(*,*) 'Could not find variable >',VarName,'<'
      CALL gdxExit(1)
    END IF

    status = gdxSymbolInfo(PGDX, VarNr, VarName, dim, VarType)
    IF (Dim /= 2 .OR. DT_VAR /= VarType) THEN
      WRITE(*,*) '**** X is not a two dimensional variable: ',Dim,':',VarType
      CALL gdxExit(1)
    END IF

    status = gdxDataReadStrStart(PGDX,VarNr,NrRecs)
    IF (status == 0) THEN
      CALL gdxerror(gdxGetLastError(PGDX), 'gdxDataReadStrStart')
    END IF
    DO WHILE (0 .ne. gdxDataReadStr(PGDX,Indx,Values,FDim))
      IF (0D0 == Values(VAL_LEVEL)) CYCLE ! skip, level 0.0 is default
      DO D = 1,Dim 
		 IF (D /= DIM) THEN
	        WRITE(*,*) Indx(D) ,'.'
		 ELSE
	        WRITE(*,*) Indx(D)
		 END IF  
	  END DO
      write(*,*) ' = ', Values(VAL_LEVEL)
	END DO
    WRITE(*,*) 'All solution values shown'
    
    status = gdxDataReadDone(PGDX)
    status = gdxGetLastError(PGDX)
    IF (status /= 0) THEN
      CALL gdxerror(status, 'GDX')
    END IF

    status = gdxClose(PGDX)
    IF (status /= 0) THEN
      CALL gdxerror(gdxGetLastError(PGDX), 'gdxClose')
    END IF

    ReadSolutionData = 0
  END FUNCTION
 
  SUBROUTINE gdxerror(i, s)
    INTEGER(KIND=4)    :: i,rc
    CHARACTER(LEN=*)   :: s
    CHARACTER(LEN=255) :: Msg
    rc = gdxErrorStr(PGDX, i, msg)
    WRITE(*,*) s, ' failed: ',msg
    CALL gdxExit(1)
  END SUBROUTINE
 
  SUBROUTINE TERMINATE
    LOGICAL :: brc
    brc = optFree(POPT)
    brc = gdxFree(PGDX)
    brc = gamsxFree(PGAX)
    CALL gdxExit(status)
  END SUBROUTINE
END PROGRAM xp_example2
